fdbserver/DataDistribution.actor.cpp

```diff
@@ -1802,54 +1801,66 @@ ACTOR Future<Void> bulkLoadJobDispatcher(Reference<DataDistributor> self,
 	return Void();
 }
 
+ACTOR Future<Void> bulkLoadJobManager(Reference<DataDistributor> self) {
+	// Find any existing bulkload job metadata. If not existing, exit.
+	state Database cx = self->txnProcessor->context();
+	Optional<BulkLoadJobState> job = wait(getAliveBulkLoadJob(cx));
+	if (!job.present()) {
+		TraceEvent(SevInfo, "DDBulkLoadJobNoJobExist", self->ddId);
+		return Void();
+	}
+	state UID jobId = job.get().getJobId();
+	state KeyRange jobRange = job.get().getJobRange();
+	state std::string jobRoot = job.get().getJobRoot();
+	state BulkLoadTransportMethod jobTransportMethod = job.get().getTransportMethod();
+	TraceEvent(SevInfo, "DDBulkLoadCoreFoundJobToRun", self->ddId)
+	    .detail("JobId", jobId)
+	    .detail("JobRange", jobRange)
+	    .detail("JobRoot", jobRoot)
+	    .detail("JobTransportMethod", jobTransportMethod);
+
+	// Set up all metadata and information required to run the job.
+	std::string localFolder = getBulkLoadJobRoot(self->bulkLoadFolder, jobId);
+	state std::string manifestLocalTempFolder = abspath(joinPath(localFolder, "manifest-temp"));
+	platform::eraseDirectoryRecursive(abspath(manifestLocalTempFolder));
+	ASSERT(platform::createDirectory(abspath(manifestLocalTempFolder)));
+	std::string remoteFolder = getBulkLoadJobRoot(jobRoot, jobId);
+	std::string jobManifestFileName = getBulkLoadJobManifestFileName();
+	std::string localJobManifestFilePath = joinPath(localFolder, jobManifestFileName);
+	std::string remoteJobManifestFilePath = appendToPath(remoteFolder, jobManifestFileName);
+	std::unordered_map<Key, BulkLoadJobFileManifestEntry> manifestEntryMap = wait(fetchBulkLoadTaskManifestEntryMap(
+	    jobTransportMethod, localJobManifestFilePath, remoteJobManifestFilePath, jobRange, self->ddId));
+	TraceEvent(SevInfo, "DDBulkLoadJobExecutorGotManifestEntryMap", self->ddId)
+	    .detail("JobId", jobId)
+	    .detail("JobRange", jobRange)
+	    .detail("JobRoot", jobRoot)
+	    .detail("JobTransportMethod", jobTransportMethod);
+	self->bulkLoadJobManager =
+	    DDBulkLoadJobManager(jobId, jobRange, jobRoot, jobTransportMethod, manifestEntryMap, manifestLocalTempFolder);
+
+	// Kick bulkload start on the job range.
+	wait(bulkLoadJobDispatcher(self, jobId, jobRange, jobRoot, jobTransportMethod));
+
+	// Wait until all bulkload job metadata on the job range is marked as complete.
+	// When completing, clear the job metadata.
+	loop {
+		bool complete = wait(checkBulkLoadJobComplete(cx, jobId));
+		if (complete) {
+			TraceEvent(SevInfo, "DDBulkLoadJobAllComplete", self->ddId).detail("JobId", jobId);
+			wait(clearBulkLoadJob(cx, jobId));
+			TraceEvent(SevInfo, "DDBulkLoadJobCleared", self->ddId).detail("JobId", jobId);
+			break;
+		}
+		wait(delay(SERVER_KNOBS->DD_BULKLOAD_SCHEDULE_MIN_INTERVAL_SEC));
```

Comment:
What happens if we crash in here? How does the bulkload registration continue?

Reply:
bulkLoadingCore is wait on this ACTOR. For any error other than actor_cancel and movekey_conflict, we retry the bulkLoadJobManager by bulkLoadingCore